// appium-ios-driver/build/test/e2e/helpers/recipes.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

function elOrNull(driver, using, value) {
  var els;
  return _regeneratorRuntime.async(function elOrNull$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(driver.findElements(using, value));

      case 2:
        els = context$1$0.sent;
        return context$1$0.abrupt('return', els[0]);

      case 4:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function clickBack(driver) {
  var el;
  return _regeneratorRuntime.async(function clickBack$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(elOrNull(driver, 'accessibility id', 'Back'));

      case 2:
        el = context$1$0.sent;
        context$1$0.t0 = el;

        if (!context$1$0.t0) {
          context$1$0.next = 8;
          break;
        }

        context$1$0.next = 7;
        return _regeneratorRuntime.awrap(driver.elementDisplayed(el));

      case 7:
        context$1$0.t0 = context$1$0.sent;

      case 8:
        if (!context$1$0.t0) {
          context$1$0.next = 11;
          break;
        }

        context$1$0.next = 11;
        return _regeneratorRuntime.awrap(driver.click(el));

      case 11:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function clickButton(driver, name) {
  var el;
  return _regeneratorRuntime.async(function clickButton$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(elOrNull(driver, 'xpath', '//UIAButton[@name = \'' + name + '\']'));

      case 2:
        el = context$1$0.sent;
        context$1$0.t0 = el;

        if (!context$1$0.t0) {
          context$1$0.next = 8;
          break;
        }

        context$1$0.next = 7;
        return _regeneratorRuntime.awrap(driver.elementDisplayed(el));

      case 7:
        context$1$0.t0 = context$1$0.sent;

      case 8:
        if (!context$1$0.t0) {
          context$1$0.next = 11;
          break;
        }

        context$1$0.next = 11;
        return _regeneratorRuntime.awrap(driver.click(el));

      case 11:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function throwMatchableError(err) {
  // When using several versions of a library (i.e. mobile-json-wire-protocol),
  // instanceof doesn't work as expected, so in tests we just use jsonwpCode
  // and message to determine that the expected error was sent.
  throw new Error('jsonwpCode: ' + err.jsonwpCode + ' ' + err.message);
}

function filterDisplayed(driver, els) {
  var displayedEls;
  return _regeneratorRuntime.async(function filterDisplayed$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(_bluebird2['default'].all(_lodash2['default'].map(els, function (el) {
          return driver.elementDisplayed(el);
        })));

      case 2:
        displayedEls = context$1$0.sent;
        return context$1$0.abrupt('return', _lodash2['default'].filter(els, function (el, i) {
          return displayedEls[i];
        }));

      case 4:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function filterVisibleUiaSelector(selector) {
  return selector.replace(/;$/, '.withPredicate("isVisible == 1");');
}

function okIfAlert(driver) {
  var text;
  return _regeneratorRuntime.async(function okIfAlert$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        text = undefined;
        context$1$0.prev = 1;
        context$1$0.next = 4;
        return _regeneratorRuntime.awrap(driver.getAlertText());

      case 4:
        text = context$1$0.sent;
        context$1$0.next = 11;
        break;

      case 7:
        context$1$0.prev = 7;
        context$1$0.t0 = context$1$0['catch'](1);

        if (!(context$1$0.t0.jsonwpCode !== 27)) {
          context$1$0.next = 11;
          break;
        }

        throw context$1$0.t0;

      case 11:
        if (!text) {
          context$1$0.next = 14;
          break;
        }

        context$1$0.next = 14;
        return _regeneratorRuntime.awrap(driver.postAcceptAlert());

      case 14:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[1, 7]]);
}

exports.clickBack = clickBack;
exports.clickButton = clickButton;
exports.elOrNull = elOrNull;
exports.throwMatchableError = throwMatchableError;
exports.filterDisplayed = filterDisplayed;
exports.filterVisibleUiaSelector = filterVisibleUiaSelector;
exports.okIfAlert = okIfAlert;

// if NoAlertOpenError (27) continue
