// appium-ios-driver/build/lib/commands/screenshot.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _Object$assign = require('babel-runtime/core-js/object/assign')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _uuidJs = require('uuid-js');

var _uuidJs2 = _interopRequireDefault(_uuidJs);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _asyncbox = require('asyncbox');

var _appiumSupport = require('appium-support');

var _appiumUiauto = require('appium-uiauto');

var _logger = require('../logger');

var _logger2 = _interopRequireDefault(_logger);

var _mobileJsonWireProtocol = require('mobile-json-wire-protocol');

var commands = {},
    helpers = {},
    extensions = {};

commands.getScreenshot = function callee$0$0() {
  var guid, shotFile, shotFolder, shotPath, takeScreenShot, data;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    var _this = this;

    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        guid = _uuidJs2['default'].create();
        shotFile = 'screenshot' + guid;
        shotFolder = _path2['default'].resolve(this.opts.tmpDir, 'appium-instruments/Run 1/');
        context$1$0.next = 5;
        return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(shotFolder));

      case 5:
        if (context$1$0.sent) {
          context$1$0.next = 8;
          break;
        }

        context$1$0.next = 8;
        return _regeneratorRuntime.awrap(_appiumSupport.fs.mkdirp(shotFolder));

      case 8:
        shotPath = _path2['default'].resolve(shotFolder, shotFile + '.png');

        _logger2['default'].debug('Taking screenshot: \'' + shotPath + '\'');

        takeScreenShot = function takeScreenShot() {
          var screenshotWaitTimeout, startMs, success;
          return _regeneratorRuntime.async(function takeScreenShot$(context$2$0) {
            while (1) switch (context$2$0.prev = context$2$0.next) {
              case 0:
                context$2$0.next = 2;
                return _regeneratorRuntime.awrap(this.uiAutoClient.sendCommand('au.capture(\'' + shotFile + '\')'));

              case 2:
                screenshotWaitTimeout = (this.opts.screenshotWaitTimeout || 10) * 1000;

                _logger2['default'].debug('Waiting ' + screenshotWaitTimeout + ' ms for screenshot to be generated.');
                startMs = Date.now();
                success = false;

              case 6:
                if (!(Date.now() - startMs < screenshotWaitTimeout)) {
                  context$2$0.next = 16;
                  break;
                }

                context$2$0.next = 9;
                return _regeneratorRuntime.awrap(_appiumSupport.fs.hasAccess(shotPath));

              case 9:
                if (!context$2$0.sent) {
                  context$2$0.next = 12;
                  break;
                }

                success = true;
                return context$2$0.abrupt('break', 16);

              case 12:
                context$2$0.next = 14;
                return _regeneratorRuntime.awrap(_bluebird2['default'].delay(300));

              case 14:
                context$2$0.next = 6;
                break;

              case 16:
                if (success) {
                  context$2$0.next = 18;
                  break;
                }

                throw new _mobileJsonWireProtocol.errors.UnknownError('Timed out waiting for screenshot file');

              case 18:
                context$2$0.next = 20;
                return _regeneratorRuntime.awrap(this.getOrientation());

              case 20:
                context$2$0.t0 = context$2$0.sent;

                if (!(context$2$0.t0 === 'LANDSCAPE')) {
                  context$2$0.next = 25;
                  break;
                }

                _logger2['default'].debug('Rotating landscape screenshot');
                context$2$0.next = 25;
                return _regeneratorRuntime.awrap(_appiumUiauto.utils.rotateImage(shotPath, -90));

              case 25:
                context$2$0.next = 27;
                return _regeneratorRuntime.awrap(_appiumSupport.fs.readFile(shotPath));

              case 27:
                return context$2$0.abrupt('return', context$2$0.sent);

              case 28:
              case 'end':
                return context$2$0.stop();
            }
          }, null, _this);
        };

        context$1$0.next = 13;
        return _regeneratorRuntime.awrap((0, _asyncbox.retry)(3, takeScreenShot));

      case 13:
        data = context$1$0.sent;
        return context$1$0.abrupt('return', new Buffer(data).toString('base64'));

      case 15:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

_Object$assign(extensions, commands, helpers);
exports.commands = commands;
exports.helpers = helpers;
exports['default'] = extensions;

// check the rotation, and rotate if necessary

// Retrying the whole screenshot process for three times.
